File: lib/libesp32/Berry/tools/coc/hash_map.cpp
/********************************************************************
** Copyright (c) 2018-2020 Guan Wenliang
** This file is part of the Berry default interpreter.
** skiars@qq.com, https://github.com/Skiars/berry
** See Copyright Notice in the LICENSE file or at
** https://github.com/Skiars/berry/blob/master/LICENSE
********************************************************************/
#include "hash_map.h"
#include "coc_string.h"

#define is_empty(entry) ((entry).next == NODE_EMPTY)

hash_map::hash_map()
{
    resize(2);
}

hash_map::hash_map(std::map<std::string, std::string> map)
{
    resize(2);
    for (auto it : map) {
        insert(it.first, it.second);
    }
}

hash_map::~hash_map()
{
}

void hash_map::resize(size_t size)
{
    entry_table bucket = m_bucket;
    m_bucket.resize(size);
    /* set all slot to empty */
    for (int i = 0; i < size; ++i) {
        m_bucket[i].next = NODE_EMPTY;
    }
    m_lastfree = size - 1;
    for (auto slot : bucket) {
        if (!is_empty(slot)) {
            insert_p(slot.key, slot.value);
        }
    }
}

hash_map::entry* hash_map::findprev(entry *list, entry *slot)
{
    int next;
    entry *prev = list;
    for (;;) {
        next = prev->next;
        if (next == NODE_NULL || &m_bucket[next] == slot) {
            break;
        }
        prev = &m_bucket[next];
    }
    if (next == NODE_NULL) {
        return NULL;
    }
    return prev;
}

int hash_map::nextfree()
{
    while (m_lastfree >= 0) {
        if (is_empty(m_bucket[m_lastfree])) {
            return int(m_lastfree);
        }
        --m_lastfree;
    }
    return -1;
}

hash_map::entry hash_map::find(const std::string &key)
{
    uint32_t hash = coc::hashcode(key);
    entry null, *slot = &m_bucket[hash % m_bucket.size()];
    if (is_empty(*slot)) {
        return null;
    }
    while (slot->key != key) {
        if (slot->next == NODE_NULL) {
            return null;
        }
        slot = &m_bucket[slot->next];
    }
    return *slot;
}

void hash_map::insert_p(const std::string &key, const std::string value)
{
    entry *slot = &m_bucket[coc::hashcode(key) % m_bucket.size()];
    if (is_empty(*slot)) { /* empty slot */
        slot->next = NODE_NULL;
    } else {
        int newidx = nextfree();
        /* get the main-slot index */
        entry *mainslot = &m_bucket[coc::hashcode(slot->key) % m_bucket.size()];
        entry *newslot = &m_bucket[newidx]; /* get a free slot index */
        if (mainslot == slot) { /* old is main slot */
            newslot->next = mainslot->next;
            mainslot->next = newidx;
            slot = newslot;
        } else { /* link to list */
            entry *prev = findprev(mainslot, slot);
            prev->next = newidx; /* link the previous node */
            *newslot = *slot; /* copy to new slot */
            slot->next = NODE_NULL;
        }
    }
    slot->key = key;
    slot->value = value;
}

void hash_map::insert(const std::string &key, const std::string value)
{
    entry slot = find(key);
    if (slot.next == NODE_EMPTY) { /* new entry */
        if (m_count >= m_bucket.size()) {
            resize(m_bucket.size() * 2);
        }
        insert_p(key, value);
        ++m_count;
    }
}

hash_map::entry hash_map::entry_modify(entry entry, int *var_count)
{
    entry.key = coc::escape_operator(entry.key);
    if (entry.value == "var") {
        entry.value = "be_const_var("
                + std::to_string(*var_count) + ")";
        ++(*var_count);
    } else {
        entry.value = "be_const_" + entry.value;
    }
    return entry;
}

hash_map::entry_table hash_map::entry_list()
{
    entry_table list;
    int var_count = 0;

    resize(m_count);
    for (auto it : m_bucket) {
        list.push_back(entry_modify(it, &var_count));
    }
    return list;
}

int hash_map::var_count()
{
    int count = 0;

    resize(m_count);
    for (auto it : m_bucket) {
        count += it.value == "var" ? 1 : 0;
    }
    return count;
}

File: lib/libesp32/berry/tools/coc/hash_map.py
import copy
from coc_string import *

# resize a list, from https://stackoverflow.com/a/8850956
def list_resize_entry(l, newsize, filling=None):                                                                                  
    if newsize > len(l):                                                                                 
        l.extend([entry() for x in range(len(l), newsize)])                                                 
    else:                                                                                                
        del l[newsize:]

class entry:
    def __init__(self):
        self.key = ""
        self.value = ""
        self.next = hash_map.NODE_EMPTY
    
    def __repr__(self):
        return f"<entry object; key='{self.key}', value='{self.value}', next={self.next}>"

class hash_map:
    NODE_EMPTY = -2
    NODE_NULL = -1

    def __init__(self, map):
        self.count = 0
        self.lastfree = 0
        self.bucket = []

        self.resize(2)
        var_count = 0
        # replace any 'var' by its slot number
        for (key, value) in map.items():
            if value == "var":
                map[key] = var_count
                var_count += 1

        for key in sorted(map.keys()):
            self.insert(key, map[key])
    
    def __repr__(self):
        return f"<hash_map object; count={self.count}, bucket={self.bucket}, lastfree={self.lastfree}>"

    def is_empty(self, ent):
        return ent.next == hash_map.NODE_EMPTY
    
    def resize(self, size):
        bucket = copy.deepcopy(self.bucket)
        list_resize_entry(self.bucket, size)
        # set all slot to empty
        for i in range(size):
            self.bucket[i].next = hash_map.NODE_EMPTY
        self.lastfree = size - 1
        for slot in bucket:
            if slot.next != hash_map.NODE_EMPTY:
                self.insert_p(slot.key, slot.value)

    def findprev(self, prev, slot):
        #print(f"findprev: prev={prev} slot={slot}")
        while True:
            next = prev.next
            if next == hash_map.NODE_NULL or self.bucket[next] == slot: break
            prev = self.bucket[next]
        if next == hash_map.NODE_NULL: return None
        return prev
    
    def nextfree(self):
        while self.lastfree >= 0:
            if self.bucket[self.lastfree].next == hash_map.NODE_EMPTY:
                return self.lastfree
            self.lastfree -= 1
        return -1
    
    def find(self, key):
        hash = hashcode(key)
        null = entry()
        slot = self.bucket[hash % len(self.bucket)]
        if slot.next == hash_map.NODE_EMPTY:
            return null
        while slot.key != key:
            if slot.next == hash_map.NODE_NULL: return null
            slot = self.bucket[slot.next]
        return slot
    
    def insert_p(self, key, value):
        slot = self.bucket[hashcode(key) % len(self.bucket)]
        #print(f"slot={slot}, index={hashcode(key) % len(self.bucket)}")
        if slot.next == hash_map.NODE_EMPTY:
            slot.next = hash_map.NODE_NULL
        else:
            newidx = self.nextfree()
            #print(f"newidx={newidx}")
            # get the main-slot index
            mainslot = self.bucket[hashcode(slot.key) % len(self.bucket)]
            newslot = self.bucket[newidx]
            #print(f"self={self}\nslot={hashcode(key) % len(self.bucket)} idx_main={hashcode(slot.key) % len(self.bucket)} newidx={newidx}")
            #print(f"insert_p: 1={mainslot} 2={newslot}")
            if mainslot == slot:
                newslot.next = mainslot.next
                mainslot.next = newidx
                slot = newslot
            else:   # link to list
                #print(f"insert_p: 1={mainslot} 2={slot}")
                prev = self.findprev(mainslot, slot)
                prev.next = newidx      # link the previous node
                # copy to new slot
                newslot.key = slot.key
                newslot.value = slot.value
                newslot.next = slot.next
                slot.next = hash_map.NODE_NULL
        slot.key = key
        slot.value = value

    def insert(self, key, value):
        slot = self.find(key)
        if slot.next == hash_map.NODE_EMPTY:
            if self.count >= len(self.bucket):
                self.resize(len(self.bucket) * 2)
            self.insert_p(key, value)
            self.count += 1
    
    #################################################################################
    # Compute entries in the hash for modules or classes
    #################################################################################
    # return a list (entiry, var_count)
    def entry_modify(self, ent):
        ent.key = escape_operator(ent.key)
        if isinstance(ent.value, int):
            ent.value = "be_const_var(" + str(ent.value) + ")"
        else:
            ent.value = "be_const_" + ent.value
        return ent
    
    #  generate the final map
    def entry_list(self):
        l = []
        
        self.resize(self.count)
        for it in self.bucket:
            l.append(self.entry_modify(it))
        return l
    
    def var_count(self):
        count = 0

        self.resize(self.count)
        for it in self.bucket:
            if isinstance(it.value, int): count += 1
        return count

if __name__ == '__main__':
    # from hash_map import *
    m = hash_map({})
    print(m)
    m.insert("foo","bar")
    print(m)
    m.insert("foo2","bar")
    print(m)
    m.insert(".","3")
    print(m)
    m.insert("foo","bar2")
    print(m)
    m.insert("+","bar2")
    print(m)
    m.insert("a","var")
    m.insert("b","var")
    m.insert("c","var")
    print()
    print(m)
    print(f"var_count={m.var_count()}")
    print(f"entry_list={m.entry_list()}")